aws/resource_aws_budgets_budget_test.go

```diff
@@ -0,0 +1,408 @@
+package aws
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strconv"
+	"testing"
+	"text/template"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/budgets"
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAWSBudget_basic(t *testing.T) {
+	name := fmt.Sprintf("test-budget-%d", acctest.RandInt())
+	configBasicDefaults := newBudgetTestConfigDefaults(name)
+	configBasicDefaults.AccountID = "012345678910"
+	configBasicUpdate := newBudgetTestConfigUpdate(name)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		CheckDestroy: func(s *terraform.State) error {
+			return testCheckBudgetDestroy(name, testAccProvider)
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: testBudgetHCLBasicUseDefaults(configBasicDefaults),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckNoResourceAttr("aws_budgets_budget.foo", "account_id"),
+					resource.TestMatchResourceAttr("aws_budgets_budget.foo", "name", regexp.MustCompile(configBasicDefaults.BudgetName)),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "budget_type", configBasicDefaults.BudgetType),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "limit_amount", configBasicDefaults.LimitAmount),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "limit_unit", configBasicDefaults.LimitUnit),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_period_start", configBasicDefaults.TimePeriodStart),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_period_end", configBasicDefaults.TimePeriodEnd),
+					resource.TestCheckResourceAttr("aws_budgets_budget.foo", "time_unit", configBasicDefaults.TimeUnit),
+					testBudgetExists(configBasicDefaults, testAccProvider),
```

Comment:
The existence testing should be first so the acceptance testing error is more clear if it cannot create the resource (and also so we can later reference the `&budget`). 😄